spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,516 @@
+  with Logging { self: DeltaSharingSource =>
+
+  val snapshot: RemoteSnapshot = deltaLog.snapshot()
+
+  override val schema: StructType = {
+    val schemaWithoutCDC = snapshot.schema
+    if (options.readChangeFeed) {
+      throw DeltaSharingErrors.CDFNotSupportedInStreaming
+    } else {
+      schemaWithoutCDC
+    }
+  }
+
+  protected var sortedFetchedFiles: Seq[IndexedFile] = Seq.empty
+
+  protected var lastGetVersionTimestamp: Long = -1
+  protected var lastQueriedTableVersion: Long = -1
+  private val QUERY_TABLE_VERSION_INTERVAL_MILLIS = 30000 // 30 seconds
+
+  protected def intervalGetTableVersion: Long = {
+    val currentTimeMillis = System.currentTimeMillis()
+    if (lastGetVersionTimestamp == -1 ||
+      (currentTimeMillis - lastGetVersionTimestamp) >= QUERY_TABLE_VERSION_INTERVAL_MILLIS) {
+      lastQueriedTableVersion = deltaLog.client.getTableVersion(deltaLog.table)
+      lastGetVersionTimestamp = currentTimeMillis
+    }
+    lastQueriedTableVersion
+  }
+
+  /**
+   * Fetch the changes from delta sharing server starting from (fromVersion, fromIndex).
+   * The start point should not be included in the result, it should be consumed in the last batch.
+   *
+   * If sortedFetchedFiles is not empty, return directly.
+   * Else, fetch file changes from the delta sharing server and store them in sortedFetchedFiles.
+   */
+  protected def getFileChanges(
+    fromVersion: Long,
+    fromIndex: Long,
+    isStartingVersion: Boolean): Unit = {
+    if (!sortedFetchedFiles.isEmpty) { return }
+
+    if (fromVersion > intervalGetTableVersion) {
+      return
+    }
+
+    def sortAddFile(f1: AddFile, f2: AddFile): Boolean = { f1.url < f2.url }
+
+    def sortAddFileForCDF(f1: AddFileForCDF, f2: AddFileForCDF): Boolean = {
+      f1.version < f2.version || (f1.version == f2.version && f1.url < f2.url)
+    }
+
+    def appendToSortedFetchedFiles(indexedFile: IndexedFile): Unit = {
+      sortedFetchedFiles = sortedFetchedFiles :+ indexedFile
+    }
+
+    if (isStartingVersion) {
+      val tableFiles = deltaLog.client.getFiles(deltaLog.table, Nil, None, Some(fromVersion), None)
+      val numFiles = tableFiles.files.size
+      tableFiles.files.sortWith(sortAddFile).zipWithIndex.foreach{
+        case (file, index) if (index > fromIndex) =>
+          appendToSortedFetchedFiles(
+            IndexedFile(fromVersion, index, file, isLast = (index + 1 == numFiles)))
+      }
+    } else {
+      // TODO(lin.zhou) return metadata for each version, and check schema read compatibility
+      val tableFiles = deltaLog.client.getFiles(deltaLog.table, fromVersion)
+      val addFiles = verifyStreamHygieneAndFilterAddFiles(tableFiles)
+      addFiles.groupBy(a => a.version).toSeq.sortWith(_._1 < _._1).foreach{
+        case (v, adds) =>
+          val numFiles = adds.size
+          adds.sortWith(sortAddFileForCDF).zipWithIndex.foreach{
+            case (add, index) if (v > fromVersion || index > fromIndex) =>
```

Comment:
v > fromVersion || (v == fromVersion && index > fromIndex) ?

Ideally v >= fromVersion right? so we can add that check as well.